File: backend_port/bounty/bountyGen.ts
/**
 * TypeScript port of scripts/bounty.gd generate_from_details (lines 224-328), for the replacement
 * backend. Line-for-line equivalent to backend_port/bounty/bounty_gen.py, which is diffed against
 * an engine-captured fixture by compare.py.
 *
 * Feed it the JSON dumped by training/tests/bounty_refdata_probe.gd (reference_data.json).
 * Do NOT feed it webclient/app/bounty_data.json: that file was written with Godot's
 * JSON.stringify default sort_keys=true and its `categories` keys are ALPHABETISED, while the
 * generator indexes categories in the GDScript source's INSERTION order.
 *
 * BigInt is used for the 64-bit PCG state on purpose. Number cannot hold it, and a float-truncated
 * multiply silently produces a different stream after a handful of draws.
 */

const MASK64 = (1n << 64n) - 1n;
const MASK32 = (1n << 32n) - 1n;
const PCG_DEFAULT_INC_64 = 1442695040888963407n; // 0x14057B7EF767814F
const PCG_MULT = 6364136223846793005n;
const MASTERY_SUFFIX = "_mastery";

/**
 * Godot String::hash — DJB2 over UNICODE CODE POINTS.
 * `for (const ch of s)` iterates code points; s.charCodeAt would split an emoji into two
 * surrogates and hash it differently from the engine.
 */
export function godotHash(s: string): number {
  let h = 5381 >>> 0;
  for (const ch of s) {
    // Math.imul keeps the 32-bit multiply exact; h*33 + cp would lose precision past 2^53.
    h = (Math.imul(h, 33) + (ch.codePointAt(0) as number)) >>> 0;
  }
  return h >>> 0;
}

/** Godot 4 RandomNumberGenerator, for `.seed = X` + randi()/randi_range(). */
export class GodotRNG {
  private state = 0n;
  private inc = 0n;

  constructor(seedValue: number | bigint) {
    this.setSeed(seedValue);
  }

  setSeed(seedValue: number | bigint): void {
    const s = BigInt(seedValue) & MASK64; // negative seeds wrap into uint64, as in GDScript
    // canonical pcg32_srandom_r(rng, initstate = s, initseq = PCG_DEFAULT_INC_64)
    this.state = 0n;
    this.inc = ((PCG_DEFAULT_INC_64 << 1n) | 1n) & MASK64;
    this.step();
    this.state = (this.state + s) & MASK64;
    this.step();
  }

  private step(): number {
    const old = this.state;
    this.state = (old * PCG_MULT + this.inc) & MASK64;
    const xorshifted = (((old >> 18n) ^ old) >> 27n) & MASK32;
    const rot = (old >> 59n) & 31n;
    const out = ((xorshifted >> rot) | (xorshifted << ((-rot) & 31n))) & MASK32;
    return Number(out);
  }

  randi(): number {
    return this.step();
  }

  /**
   * RandomNumberGenerator::randi_range, via RandomPCG::random(int, int).
   *
   * *** THE DEGENERATE CASE DOES NOT DRAW. *** `from === to` returns immediately and advances the
   * stream ZERO times. bounty.gd hits that on every square for a character with exactly one
   * archetype (bountyTypes[randiRange(0, len - 1)] with len 1) and for a universe holding exactly
   * one other character. Always drawing here reproduced only 72 of 168 engine-captured cards.
   *
   * Reversed ranges (to < from) DO draw and behave as a plain swap — also verified against the
   * engine, though bounty.gd never produces one.
   */
  randiRange(from: number, to: number): number {
    if (to < from) {
      const t = from;
      from = to;
      to = t;
    }
    if (from === to) return from;
    return (this.step() % (to - from + 1)) + from;
  }
}

export interface BountyReferenceData {
  category_order: string[];
  categories: Record<string, string[]>;
  mission_types: string[];
  char_name_list: string[];
  roster: { path_name: string; universe: number }[];
  universe_enum: Record<string, number>;
}

export type Mission = [string, string[], number];

export class BountyTables {
  readonly categoryOrder: string[];
  readonly categories: Record<string, string[]>;
  readonly missionTypes: string[];
  readonly charNameList: string[];
  readonly universeOf: Map<string, number>;
  readonly allUniverses: number[];

  constructor(raw: BountyReferenceData) {
    this.categoryOrder = raw.category_order;
    this.categories = raw.categories;
    this.missionTypes = raw.mission_types;
    this.charNameList = raw.char_name_list;
    this.universeOf = new Map(raw.roster.map((r) => [r.path_name, r.universe]));
    this.allUniverses = [...new Set(Object.values(raw.universe_enum))];
  }

  getArchetypes(pathName: string): string[] {
    return this.categoryOrder.filter((k) => this.categories[k].includes(pathName));
  }

  /** CharacterDatabase.by_universe: universe ordinal -> path_names, minus excludePath. */
  byUniverse(excludePath: string): Map<number, string[]> {
    const out = new Map<number, string[]>();
    for (const u of this.allUniverses) out.set(u, []);
    for (const name of this.charNameList) {
      if (name !== excludePath) {
        out.get(this.universeOf.get(name) as number)!.push(name);
      }
    }
    return out;
  }
}

export function generateBounty(
  tables: BountyTables,
  playerName: string,
  pathName: string,
  rerolls: string | number,
  bountyType: "unlock" | "mastery" = "unlock",
): { missions: Mission[]; seedInput: string; seedHash: number } {
  const requiresBountyTarget = bountyType === "mastery";
  let seedInput = `${playerName}${pathName}${rerolls}`;
  if (requiresBountyTarget) seedInput += MASTERY_SUFFIX;
  const seedHash = godotHash(seedInput);
  const rng = new GodotRNG(seedHash);

  const cats = tables.categories;
  let bountyTypes = tables.getArchetypes(pathName);
  // A character in no category would index bountyTypes[-1] in GDScript; bounty.gd falls back to
  // every category. No roster character hits this today, but the branch must exist.
  if (bountyTypes.length === 0) bountyTypes = [...tables.categoryOrder];

  const universeDict = tables.byUniverse(pathName);
  const sameUniverse = universeDict.get(tables.universeOf.get(pathName) as number) as string[];

  const missions: Mission[] = [];
  for (let i = 0; i < 25; i++) {
    const missionType = tables.missionTypes[rng.randiRange(0, 5)];

    if (missionType === "with") {
      const specificOdds = rng.randiRange(0, 4);
      const bountyCategory = bountyTypes[rng.randiRange(0, bountyTypes.length - 1)];
      if (specificOdds === 0) {
        const members = cats[bountyCategory];
        const target = members[rng.randiRange(0, members.length - 1)];
        missions.push(target === pathName ? ["with", [bountyCategory], 6] : ["with", [target], 4]);
      } else if (specificOdds === 1 && sameUniverse.length !== 0) {
        const target = sameUniverse[rng.randiRange(0, sameUniverse.length - 1)];
        missions.push(["with", [target], 4]);
      } else {
        missions.push(["with", [bountyCategory], 6]);
      }
    } else if (missionType === "against") {
      const specificOdds = rng.randiRange(0, 8);
      const bountyCategory = bountyTypes[rng.randiRange(0, bountyTypes.length - 1)];
      if (specificOdds === 0) {
        const members = cats[bountyCategory];
        const target = members[rng.randiRange(0, members.length - 1)];
        missions.push(
          target === pathName ? ["against", [bountyCategory], 4] : ["against", [target], 3],
        );
      } else if (specificOdds === 1 && sameUniverse.length !== 0) {
        const target = sameUniverse[rng.randiRange(0, sameUniverse.length - 1)];
        missions.push(["against", [target], 3]);
      } else {
        missions.push(["against", [bountyCategory], 4]);
      }
    } else {
      let withOdds = rng.randiRange(0, 4);
      let targetOdds = rng.randiRange(0, 8);
      const withCategory = bountyTypes[rng.randiRange(0, bountyTypes.length - 1)];
      const targetCategory = bountyTypes[rng.randiRange(0, bountyTypes.length - 1)];
      const versus: string[] = [];

      if (withOdds === 0) {
        const members = cats[withCategory];
        const target = members[rng.randiRange(0, members.length - 1)];
        if (target === pathName) {
          versus.push(withCategory);
          // bounty.gd rewrites the odds here and scores win_count from it afterwards, so this
          // self-target collision counts as a specific pick (2), not a category (1).
          withOdds = 1;
        } else {
          versus.push(target);
        }
      } else if (withOdds === 1 && sameUniverse.length !== 0) {
        versus.push(sameUniverse[rng.randiRange(0, sameUniverse.length - 1)]);
      } else {
        versus.push(withCategory);
      }

      if (targetOdds === 0) {
        const members = cats[targetCategory];
        const target = members[rng.randiRange(0, members.length - 1)];
        if (target === pathName) {
          versus.push(targetCategory);
          targetOdds = 1;
        } else {
          versus.push(target);
        }
      } else if (targetOdds === 1 && sameUniverse.length !== 0) {
        versus.push(sameUniverse[rng.randiRange(0, sameUniverse.length - 1)]);
      } else {
        versus.push(targetCategory);
      }

      let winCount = withOdds ? 2 : 1;
      winCount += targetOdds ? 1 : 0;
      missions.push(["versus", versus, winCount]);
    }
  }

  return { missions, seedInput, seedHash };
}

File: backend_port/bounty/bounty_gen.py
"""Non-Godot port of scripts/bounty.gd generate_from_details (lines 224-328).

Reads ONLY backend_port/bounty/reference_data.json (the static tables, dumped from the engine once)
plus the four inputs the live server passes: player_name, path_name, rerolls, bounty_type.
Produces the identical 25-square mission list.

Run `python compare.py` to diff this against the engine-captured fixture.

The three details a naive port gets wrong (each verified against the fixture):
  * `categories` INSERTION order is the index space of every category draw. It is NOT alphabetical,
    and webclient/app/bounty_data.json IS alphabetical — do not generate from that file.
  * every randi_range consumes a draw, including randi_range(0, 0) for a 1-element list.
  * the `specific_odds == 1` universe branch is SKIPPED (no draw) when the character is the only
    one in its universe, falling through to the plain-category else. Eight roster characters are
    in that situation today.
"""

import json
import os

from godot_primitives import GodotRNG, godot_hash

HERE = os.path.dirname(os.path.abspath(__file__))
MASTERY_SUFFIX = "_mastery"


class BountyData:
    """The static tables, loaded once."""

    def __init__(self, path=None):
        with open(path or os.path.join(HERE, "reference_data.json"), encoding="utf-8") as fh:
            raw = json.load(fh)
        self.category_order = list(raw["category_order"])
        self.categories = {k: list(raw["categories"][k]) for k in self.category_order}
        self.mission_types = list(raw["mission_types"])
        self.mastery_suffix = raw["mastery_suffix"]
        # char_name_list order == the order CharacterDatabase.by_universe appends in.
        self.char_name_list = list(raw["char_name_list"])
        self.universe_of = {r["path_name"]: r["universe"] for r in raw["roster"]}
        # Every enum value gets a bucket, including universes with no characters — matching
        # by_universe()'s pre-seeded dict (an unknown universe would KeyError in GDScript too).
        self.all_universes = sorted(set(raw["universe_enum"].values()))

    def get_archetypes(self, path_name):
        """bounty.gd get_archetypes: iterate `categories` in insertion order."""
        return [k for k in self.category_order if path_name in self.categories[k]]

    def by_universe(self, exclude_path=""):
        """CharacterDatabase.by_universe: universe ordinal -> [path_name], minus exclude_path."""
        out = {u: [] for u in self.all_universes}
        for name in self.char_name_list:
            if name != exclude_path:
                out[self.universe_of[name]].append(name)
        return out


def generate(data: BountyData, player_name: str, path_name: str, rerolls, bounty_type="unlock"):
    """Port of Bounty.generate_from_details. Returns (missions, seed_input, seed_hash)."""
    requires_bounty_target = bounty_type == "mastery"
    seed_input = "%s%s%s" % (player_name, path_name, rerolls)
    if requires_bounty_target:
        seed_input += MASTERY_SUFFIX
    seed_hash = godot_hash(seed_input)
    rng = GodotRNG(seed_hash)

    categories = data.categories
    bounty_types = data.get_archetypes(path_name)
    if not bounty_types:
        # bounty.gd's guard: a character in no category would index bounty_types[-1].
        bounty_types = list(data.category_order)

    universe_dict = data.by_universe(path_name)
    same_universe = universe_dict[data.universe_of[path_name]]

    missions = []
    for _ in range(25):
        mission_type = data.mission_types[rng.randi_range(0, 5)]
        details = [mission_type]

        if mission_type == "with":
            specific_odds = rng.randi_range(0, 4)
            bounty_category = bounty_types[rng.randi_range(0, len(bounty_types) - 1)]
            if not specific_odds:
                members = categories[bounty_category]
                specific_target = members[rng.randi_range(0, len(members) - 1)]
                if specific_target == path_name:
                    details.append([bounty_category])
                    details.append(6)
                else:
                    details.append([specific_target])
                    details.append(4)
            elif specific_odds == 1 and len(same_universe) != 0:
                specific_target = same_universe[rng.randi_range(0, len(same_universe) - 1)]
                details.append([specific_target])
                details.append(4)
            else:
                details.append([bounty_category])
                details.append(6)

        elif mission_type == "against":
            specific_odds = rng.randi_range(0, 8)
            bounty_category = bounty_types[rng.randi_range(0, len(bounty_types) - 1)]
            if not specific_odds:
                members = categories[bounty_category]
                specific_target = members[rng.randi_range(0, len(members) - 1)]
                if specific_target == path_name:
                    details.append([bounty_category])
                    details.append(4)
                else:
                    details.append([specific_target])
                    details.append(3)
            elif specific_odds == 1 and len(same_universe) != 0:
                specific_target = same_universe[rng.randi_range(0, len(same_universe) - 1)]
                details.append([specific_target])
                details.append(3)
            else:
                details.append([bounty_category])
                details.append(4)

        else:  # "versus"
            with_specific_odds = rng.randi_range(0, 4)
            target_specific_odds = rng.randi_range(0, 8)
            with_bounty_category = bounty_types[rng.randi_range(0, len(bounty_types) - 1)]
            target_bounty_category = bounty_types[rng.randi_range(0, len(bounty_types) - 1)]
            versus_details = []

            if not with_specific_odds:
                members = categories[with_bounty_category]
                specific_target = members[rng.randi_range(0, len(members) - 1)]
                if specific_target == path_name:
                    versus_details.append(with_bounty_category)
                    # NOTE: the GDScript rewrites the odds variable here, and win_count is
                    # computed from it AFTERWARDS — so this self-target collision is scored as
                    # if it had been a specific pick (2 points), not as a category (1).
                    with_specific_odds = 1
                else:
                    versus_details.append(specific_target)
            elif with_specific_odds == 1 and len(same_universe) != 0:
                specific_target = same_universe[rng.randi_range(0, len(same_universe) - 1)]
                versus_details.append(specific_target)
            else:
                versus_details.append(with_bounty_category)

            if not target_specific_odds:
                members = categories[target_bounty_category]
                specific_target = members[rng.randi_range(0, len(members) - 1)]
                if specific_target == path_name:
                    versus_details.append(target_bounty_category)
                    target_specific_odds = 1
                else:
                    versus_details.append(specific_target)
            elif target_specific_odds == 1 and len(same_universe) != 0:
                specific_target = same_universe[rng.randi_range(0, len(same_universe) - 1)]
                versus_details.append(specific_target)
            else:
                versus_details.append(target_bounty_category)

            details.append(versus_details)
            win_count = 2 if with_specific_odds else 1
            win_count += 1 if target_specific_odds else 0
            details.append(win_count)

        missions.append(details)

    return missions, seed_input, seed_hash
